src/vsr/sync.zig

```diff
@@ -0,0 +1,288 @@
+    cancel_commit,
+
+    /// Waiting for `Grid.cancel()`.
+    cancel_grid,
+
+    /// We need to sync, but are waiting for a usable `sync_target_max`.
+    request_target,
+
+    request_trailers: struct {
+        target: SyncTarget,
+        manifest: SyncTrailer = .{},
+        free_set: SyncTrailer = .{},
+        client_sessions: SyncTrailer = .{},
+        /// Our new VSRState.previous_checkpoint_id.
+        /// Arrives with command=sync_free_set.
+        previous_checkpoint_id: ?u128 = null,
+        /// The checksum of the prepare corresponding to checkpoint_op.
+        /// Arrives with command=sync_client_sessions.
+        checkpoint_op_checksum: ?u128 = null,
+
+        pub fn done(self: *const @This()) bool {
+            if (self.free_set.done) assert(self.previous_checkpoint_id != null);
+            if (self.client_sessions.done) assert(self.checkpoint_op_checksum != null);
+
+            return self.manifest.done and self.free_set.done and self.client_sessions.done;
+        }
+    },
+
+    superblock_update: struct {
+        target: SyncTarget,
+        previous_checkpoint_id: u128,
+        checkpoint_op_checksum: u128,
+    },
+
+    done,
+
+    pub fn valid_transition(from: std.meta.Tag(SyncStage), to: std.meta.Tag(SyncStage)) bool {
+        return switch (from) {
+            .none => to == .cancel_commit or
+                to == .cancel_grid or
+                to == .request_target,
+            .cancel_commit => to == .cancel_grid,
+            .cancel_grid => to == .request_target,
+            .request_target => to == .request_target or
+                to == .request_trailers,
+            .request_trailers => to == .request_trailers or
+                to == .superblock_update,
+            .superblock_update => to == .request_trailers or
+                to == .done,
+            .done => to == .request_trailers or
+                to == .none,
+        };
+    }
+
+    pub fn target(stage: *const SyncStage) ?SyncTarget {
+        return switch (stage.*) {
+            .none,
+            .cancel_commit,
+            .cancel_grid,
+            .request_target,
+            .done,
+            => null,
+            .request_trailers => |s| s.target,
+            .superblock_update => |s| s.target,
+        };
+    }
+};
+
+pub const SyncTargetCandidate = struct {
+    /// The target's checkpoint identifier.
+    checkpoint_id: u128,
+    /// The op_checkpoint() that corresponds to the checkpoint id.
+    checkpoint_op: u64,
+
+    pub fn canonical(target: SyncTargetCandidate) SyncTarget {
+        return .{
+            .checkpoint_id = target.checkpoint_id,
+            .checkpoint_op = target.checkpoint_op,
+        };
+    }
+};
+
+pub const SyncTarget = struct {
+    /// The target's checkpoint identifier.
+    checkpoint_id: u128,
+    /// The op_checkpoint() that corresponds to the checkpoint id.
+    checkpoint_op: u64,
+};
+
+pub const SyncTargetQuorum = struct {
```

Comment:
For these types (and others), within this file, could we drop the `Sync` prefix (we can keep that outside this file)?

And then, also drop `Target` from SyncTargetQuorum?

Then for example, the signature for `replace()` will read like: `pub fn replace(quorum: *Quorum...`.

Reply:
This will remove repetition and help make the control flow _pop_.

Reply:
For example, the `candidates` field would then look like:

```zig
candidates: [constants.replicas_max]?Target = [_]?Target{null} ** constants.replicas_max
```

... helping to make the type attributes themselves clearer.

Reply:
I removed the `Sync` prefix (mixed feelings here though, since it makes the source a little less grep-able).

I left `Target` in `TargetQuorum` though; I think it is important. "sync quorum" doesn't really make sense to me, "sync target quorum" is immediately clear.